rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,106 @@
+# Proposal to add a reusable kernel knob
+
+## Introduction
+
+Primitive creation is an expensive operation, which is why oneDNN has a
+primitive cache, a persistent cache, and a kernel cache (for the GPU). For GPU
+implementations, the kernel cache is optional and must be opted into. Although
+compiling kernels for a given problem shape can open up optimization potential,
+it comes at the cost of being incompatible with the kernel cache because the
+kernel cannot be reused for different problem shapes. Therefore, it can be
+helpful to the user to be able to specify if they prefer to use a kernel that
+has been designed to be reusable via the kernel cache (accepting any possible
+performance impacts associated with it) or the typical performance-optimized
+kernel. This proposal is to add a knob the user can control to specify whether
+they would like reusable kernels to be dispatched preferentially over the
+standard performance-optimized kernel ordering for GPU implementations.
+
+> **Note:** Reusable kernels are defined as GPU kernels with a finite set of
```

Comment:
With post-ops we can have reusability only when the number of used post-op chains is reasonably limited. Otherwise in general kernels are not reusable.

Do you have anything more specific in mind for the formal definition of "reusable kernels"? As a criteria, I thought we can say that some functionality is covered with reusable kernels only when they can be potentially converted to AOT kernels hence we need a fairly limited set of unique kernels. With post-ops it's not the case so we can't cover them with reusable kernels given the current implement approach.

Reply:
To me, a reusable kernel is a specialization intended for use on a "large/notable" set of inputs. For example, a kernel with post-ops may be reused when the same post-ops are applied. This kind of support is still beneficial for dynamic workloads where some of the input data dimensions are changing, but the core operation is the same, such as can occur with `matmul+eltwise` fusion. While AOT compiled kernels must be reusable (and additionally meet memory size constraints), this is a strict subset of what I think we need to support.